File: desktop-tauri/src-tauri/src/commands/review.rs
use std::path::PathBuf;

use vibyra_core::review::{self, MergeOutcome, PruneOutcome, WorktreeEntry, WorktreeStatus};
use vibyra_core::CoreError;

use super::run_blocking_core;

// The Review tool's native half. Every command takes the safe-workspace ref
// the pane has carried since launch (`SessionInfo.workspace`) rather than a
// session id, so a suspended pane — whose process is gone — reviews exactly
// like a live one. The `vibyra/*` branch guard inside `review` is what keeps
// these from acting on any other folder.

#[tauri::command]
pub async fn review_status(
    worktree: String,
    base_commit: String,
) -> Result<WorktreeStatus, CoreError> {
    run_blocking_core(move || review::worktree_status(&PathBuf::from(worktree), &base_commit)).await
}

#[tauri::command]
pub async fn review_file_diff(
    worktree: String,
    base_commit: String,
    path: String,
) -> Result<String, CoreError> {
    run_blocking_core(move || review::file_diff(&PathBuf::from(worktree), &base_commit, &path))
        .await
}

/// `paths` selects which of the worktree's files to land, repo-root-relative.
/// Optional rather than required, and absent means all of them, so a caller
/// with no selection to make keeps calling this exactly as it always has.
#[tauri::command]
pub async fn review_merge(
    project_root: String,
    worktree: String,
    base_commit: String,
    paths: Option<Vec<String>>,
) -> Result<MergeOutcome, CoreError> {
    run_blocking_core(move || {
        review::merge_back(
            &PathBuf::from(project_root),
            &PathBuf::from(worktree),
            &base_commit,
            &paths.unwrap_or_default(),
        )
    })
    .await
}

#[tauri::command]
pub async fn review_discard(project_root: String, worktree: String) -> Result<(), CoreError> {
    run_blocking_core(move || {
        review::discard_worktree(&PathBuf::from(project_root), &PathBuf::from(worktree))
    })
    .await
}

/// Housekeeping. Both are guarded to `vibyra/*` inside the core, so neither
/// can see or touch a worktree or branch the app did not create.
#[tauri::command]
pub async fn review_list_worktrees(project_root: String) -> Result<Vec<WorktreeEntry>, CoreError> {
    run_blocking_core(move || review::list_worktrees(&PathBuf::from(project_root))).await
}

#[tauri::command]
pub async fn review_prune_worktrees(project_root: String) -> Result<PruneOutcome, CoreError> {
    run_blocking_core(move || review::prune_worktrees(&PathBuf::from(project_root))).await
}

File: desktop-tauri/src/ipc/review.ts
import { invoke } from "@tauri-apps/api/core";

import type { SafeWorkspaceRef } from "../types";

export type ChangeKind = "added" | "modified" | "deleted" | "renamed";

export interface ChangedFile {
  path: string;
  kind: ChangeKind;
  additions: number;
  deletions: number;
}

export interface WorktreeStatus {
  changed: ChangedFile[];
  /** True when the list was cut at the native ceiling. */
  truncated: boolean;
}

export interface MergeOutcome {
  applied: boolean;
  /** Files that stopped the merge; non-empty means nothing was changed. */
  conflicts: string[];
}

export function reviewStatus(workspace: SafeWorkspaceRef): Promise<WorktreeStatus> {
  return invoke("review_status", {
    worktree: workspace.path,
    baseCommit: workspace.baseCommit,
  });
}

export function reviewFileDiff(workspace: SafeWorkspaceRef, path: string): Promise<string> {
  return invoke("review_file_diff", {
    worktree: workspace.path,
    baseCommit: workspace.baseCommit,
    path,
  });
}

/**
 * Lands the workspace's changes in the project as ordinary working-tree edits.
 *
 * `paths` narrows that to a selection — the same repo-root-relative strings
 * `reviewStatus` returns. Omitting it means everything, which is what the
 * whole-workspace actions send; the native side treats an absent list and an
 * empty one alike. All-or-nothing holds over whatever was selected.
 */
export function reviewMerge(
  projectRoot: string,
  workspace: SafeWorkspaceRef,
  paths?: string[],
): Promise<MergeOutcome> {
  return invoke("review_merge", {
    projectRoot,
    worktree: workspace.path,
    baseCommit: workspace.baseCommit,
    paths,
  });
}

export function reviewDiscard(projectRoot: string, workspace: SafeWorkspaceRef): Promise<void> {
  return invoke("review_discard", { projectRoot, worktree: workspace.path });
}

/** A `vibyra/*` worktree on disk, whether or not a pane still owns it. */
export interface WorktreeEntry {
  path: string;
  branch: string;
  head: string;
  locked: boolean;
  /** False once the folder is gone but git has not pruned the record yet. */
  exists: boolean;
}

export interface PruneOutcome {
  prunedWorktrees: number;
  deletedBranches: string[];
  sweptFiles: number;
}

/**
 * Every safe-mode worktree this project has, so the fleet can show the ones
 * no pane owns. Closing a pane with its X — rather than through Discard —
 * leaves the worktree behind, and until it is listed here it is invisible.
 */
export function reviewListWorktrees(projectRoot: string): Promise<WorktreeEntry[]> {
  return invoke("review_list_worktrees", { projectRoot });
}

/** Reaps worktrees git has lost, their merged branches, and scratch files. */
export function reviewPruneWorktrees(projectRoot: string): Promise<PruneOutcome> {
  return invoke("review_prune_worktrees", { projectRoot });
}
